### PR TITLE
abci.md fixup

### PR DESCRIPTION
**Pull request by williambanfield**
This pull requests offers several fixups to the `abci.md` document. The changes are primarily in areas that I found somewhat unclear.

### CODE REVIEW

File: spec/abci/abci.md [hunk 1]
```diff
@@ -610,15 +621,15 @@ via light client.
 
 ## Data Types
 
-The data types not listed below are the same as the [core data structures](../core/data_structures.md). The ones listed below have specific changes to better accommodate applications.
+The data types not listed below are the same as the [core data structures](../spec/core/data_structures.md). The ones listed below have specific changes to better accommodate applications.
```
Comment (creachadair) [inline spec/abci/abci.md:624]: This seems confusing, I'm not sure how the reader is expected to know which types are supposed to be here vs. not here ("anyone who is absent, please raise your hand"). 

What I _think_ this is saying is something like:

> Most of the data structures used in ABCI are shared [common data structures](../spec/core/data_structures.md) with Tendermint Core. In certain cases, however, ABCI uses different structures, which are documented here:

**Review by creachadair — COMMENTED** [review]
A bunch of relatively minor questions and suggestions. Where I've made suggestions, please treat them as examples rather than strict replacements, since I may have gotten things wrong.

File: spec/abci/abci.md [hunk 2]
```diff
@@ -45,33 +45,33 @@ More details on managing state across connections can be found in the section on
...
-during the method's execution.
+The `CheckTx`, `BeginBlock`, `DeliverTx`, `EndBlock` methods include an `Events`
+field in their `Response*`. Each event contains a `type` and a list of `EventAttribute`s,
+which are key-value pairs denoting something about what happened during the method's execution.
 
-Events can be used to index transactions and blocks according to what happened
+`Event`s can be used to index transactions and blocks according to what happened
 during their execution. Note that the set of events returned for a block from
 `BeginBlock` and `EndBlock` are merged. In case both methods return the same
-tag, only the value defined in `EndBlock` is used.
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply spec/abci/abci.md:71]: This is a great question that I do not know the answer to. I had assumed that, based on the statement on line 71, that the merging was really an insertion into the set that overwrote the previous value. Before I go spelunking through the event indexing code to find out, perhaps @alexanderbez may know the answer here?

File: spec/abci/abci.md [hunk 3]
```diff
@@ -355,16 +363,16 @@ via light client.
...
-    any DeliverTxs.
+    * Signals the beginning of a new block.
+    * Called prior to any `DeliverTx` method calls.
     * The header contains the height, timestamp, and more - it exactly matches the
     Tendermint block header. We may seek to generalize this in the future.
     * The `LastCommitInfo` and `ByzantineValidators` can be used to determine
-    rewards and punishments for the validators. NOTE validators here do not
-    include pubkeys.
+    rewards and punishments for the validators. **NOTE:** validators here do not
+    include public keys.
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply spec/abci/abci.md]: Removed.

File: spec/abci/abci.md [hunk 4]
```diff
@@ -393,10 +402,10 @@ via light client.
...
 
     * Technically optional - not involved in processing blocks.
-    * Guardian of the mempool: every node runs CheckTx before letting a
+    * Guardian of the mempool: every node runs `CheckTx` before letting a
     transaction into its local mempool.
     * The transaction may come from an external user or another node
-    * CheckTx need not execute the transaction in full, but rather a light-weight
+    * `CheckTx` need not execute the transaction in full, but rather a light-weight
     yet stateful validation, like checking signatures and account balances, but
     not running code in a virtual machine.
```
Comment (creachadair) [inline spec/abci/abci.md:410]: It need not, but I think we perhaps want a stronger statement: We seem to _rely_ on the fact that checking a transaction doesn't actually apply any state changes. If that's correct, then "stateful" here may be doing a bit more work than it should. What I think this is trying to say is something like

> `CheckTx` validates the transaction against the current state of the application, for example checking signatures and account balances, but does not apply any state changes described in the transaction.

Does that seem right? Presumably the application could actually run the transaction against a non-persistent clone of the current state to see if it's valid (an implementation detail) but it seems like "non-persistent" is important.

**Review by creachadair — COMMENTED** [review]
This is looking pretty good. I have a few more questions, and as usual some cosmetic suggestions.

File: spec/abci/abci.md [hunk 5]
```diff
@@ -45,33 +45,34 @@ More details on managing state across connections can be found in the section on
...
-Some methods (`CheckTx, BeginBlock, DeliverTx, EndBlock`)
-include an `Events` field in their `Response*`. Each event contains a type and a
-list of attributes, which are key-value pairs denoting something about what happened
-during the method's execution.
+The `CheckTx`, `BeginBlock`, `DeliverTx`, `EndBlock` methods include an `Events`
+field in their `Response*`. Each event contains a `type` and a list of `EventAttribute`s,
+which are key-value pairs denoting something about what happened during the method's execution.
 
-Events can be used to index transactions and blocks according to what happened
+`Event` values can be used to index transactions and blocks according to what happened
```
Comment (cmwaters) [inline spec/abci/abci.md:69]: events

I find adding a schematic to the descriptions will help with the more visual learners

File: spec/abci/abci.md [hunk 6]
```diff
@@ -130,11 +133,13 @@ Example:
...
 
-A part of Tendermint's security model is the use of evidence which serves as proof of
+Tendermint's security model relies on the use of "evidence". Evidence is proof of
 malicious behaviour by a network participant. It is the responsibility of Tendermint
-to detect such malicious behaviour, to gossip this and commit it to the chain and once
-verified by all validators to pass it on to the application through the ABCI. It is the
-responsibility of the application then to handle the evidence and exercise punishment.
+to detect such malicious behaviour. When malicious behavior is detected, Tendermint
+will gossip evidence of the behavior to other nodes and commit the evidence to 
+the chain and once it is verified by all validators. This evidence will then be
```
Comment (cmwaters) [inline spec/abci/abci.md:140]:
```suggestion
the chain once it is verified by all validators. This evidence will then be 
```

File: spec/abci/abci.md [hunk 7]
```diff
@@ -214,17 +219,22 @@ Commit are included in the header of the next block.
...
-[state sync section](apps.md#state-sync).
+[state sync section](../spec/p2p/messages/state-sync.md).
 
-When a new node is discovering snapshots in the P2P network, existing nodes will call
-`ListSnapshots` on the application to retrieve any local state snapshots. The new node will
-offer these snapshots to its local application via `OfferSnapshot`.
+New nodes will discover and request snapshots from other nodes in the P2P network.
+A Tendermint node that receives a request for snapshots from a peer will call
+`ListSnapshots` on its application to retrieve any local state snapshots. In addition,
+the new node will offer each snapshot received from a peer to its local application
```
Comment (cmwaters) [inline spec/abci/abci.md:227]:
```suggestion
`ListSnapshots` on its application to retrieve any local state snapshots. After receiving snapshots from peers,
the new node will offer each snapshot to its local application
```

**Review by cmwaters — APPROVED** [review]
LGTM

File: spec/abci/abci.md [hunk 8]
```diff
@@ -45,33 +45,34 @@ More details on managing state across connections can be found in the section on
...
+
 ## Events
 
-Some methods (`CheckTx, BeginBlock, DeliverTx, EndBlock`)
-include an `Events` field in their `Response*`. Each event contains a type and a
-list of attributes, which are key-value pairs denoting something about what happened
-during the method's execution.
+The `CheckTx`, `BeginBlock`, `DeliverTx`, `EndBlock` methods include an `Events`
+field in their `Response*`. Each event contains a `type` and a list of `EventAttribute`s,
+which are key-value pairs denoting something about what happened during the method's execution.
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply spec/abci/abci.md]: There does appear to be this separate page for application development that exists within the [docs folder](https://docs.tendermint.com/master/app-dev/indexing-transactions.html) within tendermint as well as this page on [subscribing to events](https://docs.tendermint.com/master/tendermint-core/subscription.html). None of these seem to exactly define 'what' an event is or how they should be used. I'll add a note that these do not impact consensus in any way here.

File: spec/abci/abci.md [hunk 9]
```diff
@@ -45,33 +45,34 @@ More details on managing state across connections can be found in the section on
...
 during their execution. Note that the set of events returned for a block from
 `BeginBlock` and `EndBlock` are merged. In case both methods return the same
-tag, only the value defined in `EndBlock` is used.
+key, only the value defined in `EndBlock` is used.
 
 Each event has a `type` which is meant to categorize the event for a particular
-`Response*` or tx. A `Response*` or tx may contain multiple events with duplicate
+`Response*` or `Tx`. A `Response*` or `Tx` may contain multiple events with duplicate
 `type` values, where each distinct entry is meant to categorize attributes for a
 particular event. Every key and value in an event's attributes must be UTF-8
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply spec/abci/abci.md:104]: @marbar3778  Do you know the answer to this?

File: spec/abci/abci.md (hunk 5 above)
Comment (williambanfield) [reply spec/abci/abci.md:96]: I think it's not a bad call, but I'd prefer to may do a bit more polishing if adding an image. My goal with this PR was to spot-fix a few issues or ambiguities I noticed in the spec as it related to issue #316.

File: spec/abci/abci.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (williambanfield) [reply spec/abci/abci.md:71]: https://github.com/tendermint/spec/issues/345

File: spec/abci/abci.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (williambanfield) [reply spec/abci/abci.md:104]: https://github.com/tendermint/spec/issues/346

**Pull request merged**